XiEditor/StyleMap.swift

```diff
@@ -61,15 +57,46 @@ func utf8_offset_to_utf16(_ s: String, _ ix: Int) -> Int {
 class StyleMap {
     private var map: [Style?] = []
 
-    func defStyle(json: [String: AnyObject]) {
+    func defStyle(for fromFont: NSFont, json: [String: AnyObject]) {
         guard let styleID = json["id"] as? Int else { return }
-        let fgColor = json["fg_color"] as? UInt32 ?? 0xFF000000
-        let bgColor = json["bg_color"] as? UInt32 ?? 0
-        let weight = json["weight"] as? UInt16 ?? 400
-        let underline = json["underline"] as? Bool ?? false
+
+        var attrs: [String: Any] = [
+            NSForegroundColorAttributeName: json["fg_color"] as? NSColor ?? NSColor.red,
+        ]
+
+        if let bgColor = json["bg_color"] as? NSColor, bgColor.alphaComponent != 0.0 {
+            attrs[NSBackgroundColorAttributeName] = bgColor
+        }
+
+        if let underline = json["underline"] as? Bool, underline {
+            attrs[NSUnderlineStyleAttributeName] = NSUnderlineStyle.styleSingle.rawValue
+        }
+
+        let fm = NSFontManager.shared()
+        var font: NSFont?
+
         let italic = json["italic"] as? Bool ?? false
+        // convert to NSFont weight, which is a value between 1-15 with 5 being the normal weight,
+        // from CSS-like values that are between 100-900
+        let weight = (json["weight"] as? Int ?? fm.weight(of: fromFont)) / 100 + 1
+
+        if italic {
```

Comment:
I think the `fm.weight(of:)` method requires a slightly different conversion. https://bugs.webkit.org/show_bug.cgi?id=6484 has the info but I'll give the table below:

    100 -> 2
    200 -> 3
    300 -> 4
    400 -> 5
    500 -> 6
    600 -> 8
    700 -> 9
    800 -> 10
    900 -> 12

I heard somewhere (not sure where) that 1000 is or will be a valid CSS value. I think it should map to 15.